File: ports/khinchin.go
// Khinchin's constant (OEIS A002210) via the accelerated zeta series.
//
// Same mathematics as ../khinchin_fast.c:
//
//	ln(2) ln(K0) = -sum_{k=2}^{N-1} ln((k-1)/k) ln((k+1)/k)
//	               + sum_{n>=1} (zeta(2n) - 1 - sum_{k=2}^{N-1} k^(-2n)) / n * h(n),
//
//	h(n) = sum_{j=1}^{2n-1} (-1)^(j+1)/j.
//
// Runs entirely on math/big.Int in fixed point (scale 2^wp) — the
// "native bignum" data point for Go, no dependencies. math/big has no
// transcendentals, so the port carries its own fixed-point kit: pi by
// Machin's formula, logarithms via the atanh(1/q) series, exp by
// argument-halving plus Taylor. The even zeta values come from the
// positive-term recurrence (n + 1/2) zeta(2n) = sum_j zeta(2j)
// zeta(2n-2j) — O(M^2) bignum products, with the convolutions (halved
// via their j <-> n-j symmetry) fanned out over goroutines. Guard bits
// absorb the drift; nothing here is interval-certified.
//
// Run:  go run khinchin.go DIGITS [OUTPUT_FILE]        (default 100)
//
// With OUTPUT_FILE the decimal value plus a newline is written there
// (the same contract as ../khinchin_fast.c); otherwise stdout.
package main

import (
	"fmt"
	"math"
	"math/big"
	"os"
	"runtime"
	"strconv"
	"sync"
)

var (
	wp  uint
	one *big.Int
)

func mulShift(a, b *big.Int) *big.Int {
	return new(big.Int).Rsh(new(big.Int).Mul(a, b), wp)
}

// atanh(1/q) = sum_{j>=0} 1 / ((2j+1) q^(2j+1)); atan(1/q) alternates.
func arcRecip(q int64, alternating bool) *big.Int {
	q2 := big.NewInt(q * q)
	power := new(big.Int).Div(one, big.NewInt(q))
	s := new(big.Int).Set(power)
	term := new(big.Int)
	for j := int64(1); power.Sign() > 0; j++ {
		power.Div(power, q2)
		term.Div(power, big.NewInt(2*j+1))
		if alternating && j%2 == 1 {
			s.Sub(s, term)
		} else {
			s.Add(s, term)
		}
	}
	return s
}

// exp(y) for 0 <= y < 1: halve `shift` times, Taylor, square back.
func expFixed(y *big.Int) *big.Int {
	const shift = 16
	arg := new(big.Int).Rsh(y, shift)
	acc := new(big.Int).Set(one)
	term := new(big.Int).Set(one)
	for j := int64(1); term.Sign() > 0; j++ {
		term = mulShift(term, arg)
		term.Div(term, big.NewInt(j))
		acc.Add(acc, term)
	}
	for i := 0; i < shift; i++ {
		acc = mulShift(acc, acc)
	}
	return acc
}

// khinchinFixed returns K0 scaled by 2^wp.
func khinchinFixed(digits int) *big.Int {
	prec := uint(math.Ceil(float64(digits+2)*math.Log2(10))) + 32
	wp = prec + uint(math.Ceil(math.Sqrt(float64(prec)))) + 64
	one = new(big.Int).Lsh(big.NewInt(1), wp)
	bigN := int(math.Max(3, math.Floor(math.Pow(float64(wp), 0.35))))
	m := int(math.Ceil(float64(wp)*math.Ln2/(2*math.Log(float64(bigN))))) + 1

	// Finite logarithmic correction:
	// -ln((k-1)/k) ln((k+1)/k) = 4 atanh(1/(2k-1)) atanh(1/(2k+1)).
	s := new(big.Int)
	for k := int64(2); k < int64(bigN); k++ {
		p := mulShift(arcRecip(2*k-1, false), arcRecip(2*k+1, false))
		s.Add(s, p.Lsh(p, 2))
	}

	// pi = 16 atan(1/5) - 4 atan(1/239); zeta(2) = pi^2/6 seeds the
	// positive-term recurrence (n + 1/2) z_n = sum z_j z_{n-j}, whose
	// convolutions run in parallel over goroutines.
	pi := new(big.Int).Sub(
		new(big.Int).Lsh(arcRecip(5, true), 4),
		new(big.Int).Lsh(arcRecip(239, true), 2))
	workers := runtime.NumCPU()
	z := make([]*big.Int, m+1)
	z[1] = new(big.Int).Div(mulShift(pi, pi), big.NewInt(6))
	for n := 2; n <= m; n++ {
		half := (n - 1) / 2
		raw := new(big.Int)
		if half >= 64 {
			partials := make([]*big.Int, workers)
			var wg sync.WaitGroup
			for w := 0; w < workers; w++ {
				wg.Add(1)
				go func(w int) {
					defer wg.Done()
					local := new(big.Int)
					product := new(big.Int)
					for j := 1 + half*w/workers; j <= half*(w+1)/workers; j++ {
						local.Add(local, product.Mul(z[j], z[n-j]))
					}
					partials[w] = local
				}(w)
			}
			wg.Wait()
			for _, p := range partials {
				raw.Add(raw, p)
			}
		} else {
			product := new(big.Int)
			for j := 1; j <= half; j++ {
				raw.Add(raw, product.Mul(z[j], z[n-j]))
			}
		}
		acc := new(big.Int).Lsh(new(big.Int).Rsh(raw, wp), 1)
		if n%2 == 0 {
			acc.Add(acc, mulShift(z[n/2], z[n/2]))
		}
		z[n] = acc.Div(acc.Lsh(acc, 1), big.NewInt(int64(2*n+1)))
	}

	// Main accelerated loop, ascending, incremental power table.
	powers := make([]*big.Int, 0, bigN-2)
	for k := int64(2); k < int64(bigN); k++ {
		powers = append(powers, new(big.Int).Div(one, big.NewInt(k*k)))
	}
	h := new(big.Int).Set(one)
	tail := new(big.Int)
	for n := int64(1); n <= int64(m); n++ {
		tail.Sub(z[n], one)
		for _, p := range powers {
			tail.Sub(tail, p)
		}
		term := new(big.Int).Mul(tail, h)
		term.Div(term, big.NewInt(n))
		s.Add(s, term.Rsh(term, wp))
		h.Add(h, new(big.Int).Div(one, big.NewInt(2*n+1)))
		h.Sub(h, new(big.Int).Div(one, big.NewInt(2*n)))
		for i, k := 0, int64(2); k < int64(bigN); i, k = i+1, k+1 {
			powers[i].Div(powers[i], big.NewInt(k*k))
		}
	}

	ln2 := new(big.Int).Lsh(arcRecip(3, false), 1)
	y := new(big.Int).Div(new(big.Int).Lsh(s, wp), ln2)
	return expFixed(y)
}

func main() {
	digits := 100
	if len(os.Args) > 1 {
		digits, _ = strconv.Atoi(os.Args[1])
	}
	value := khinchinFixed(digits)
	scaled := new(big.Int).Mul(value, new(big.Int).Exp(
		big.NewInt(10), big.NewInt(int64(digits)), nil))
	scaled.Add(scaled, new(big.Int).Lsh(big.NewInt(1), wp-1))
	scaled.Rsh(scaled, wp)
	text := scaled.String()
	result := text[:1] + "." + text[1:]
	if len(os.Args) > 2 {
		if err := os.WriteFile(os.Args[2], []byte(result+"\n"), 0o644); err != nil {
			panic(err)
		}
	} else {
		fmt.Println(result)
	}
}

File: ports/Khinchin.java
// Khinchin's constant (OEIS A002210) via the accelerated zeta series.
//
// Same mathematics as ../khinchin_fast.c:
//
//   ln(2) ln(K0) = -sum_{k=2}^{N-1} ln((k-1)/k) ln((k+1)/k)
//                  + sum_{n>=1} (zeta(2n) - 1 - sum_{k=2}^{N-1} k^(-2n)) / n * h(n),
//
//   h(n) = sum_{j=1}^{2n-1} (-1)^(j+1)/j.
//
// Runs entirely on java.math.BigInteger in fixed point (scale 2^wp) —
// the "native bignum" data point for Java, no dependencies. BigInteger
// has no transcendentals, so the port carries its own fixed-point kit:
// pi by Machin's formula, logarithms via the atanh(1/q) series, exp by
// argument-halving plus Taylor. The even zeta values come from the
// positive-term recurrence (n + 1/2) zeta(2n) = sum_j zeta(2j)
// zeta(2n-2j) — O(M^2) bignum products, with the convolutions (halved
// via their j <-> n-j symmetry) run through parallel streams. Guard
// bits absorb the drift; nothing here is interval-certified.
//
// Run:  java Khinchin.java DIGITS [OUTPUT_FILE]        (default 100)
//
// With OUTPUT_FILE the decimal value plus a newline is written there
// (the same contract as ../khinchin_fast.c); otherwise stdout.

import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.IntStream;

public final class Khinchin {
    static int wp;
    static BigInteger one;

    static BigInteger mulShift(BigInteger a, BigInteger b) {
        return a.multiply(b).shiftRight(wp);
    }

    // atanh(1/q) = sum_{j>=0} 1/((2j+1) q^(2j+1)); atan(1/q) alternates.
    static BigInteger arcRecip(long q, boolean alternating) {
        BigInteger q2 = BigInteger.valueOf(q * q);
        BigInteger power = one.divide(BigInteger.valueOf(q));
        BigInteger s = power;
        for (long j = 1; power.signum() > 0; j++) {
            power = power.divide(q2);
            BigInteger term = power.divide(BigInteger.valueOf(2 * j + 1));
            s = (alternating && j % 2 == 1) ? s.subtract(term) : s.add(term);
        }
        return s;
    }

    // exp(y) for 0 <= y < 1: halve 16 times, Taylor, square back.
    static BigInteger expFixed(BigInteger y) {
        BigInteger arg = y.shiftRight(16);
        BigInteger acc = one, term = one;
        for (long j = 1; term.signum() > 0; j++) {
            term = mulShift(term, arg).divide(BigInteger.valueOf(j));
            acc = acc.add(term);
        }
        for (int i = 0; i < 16; i++) {
            acc = mulShift(acc, acc);
        }
        return acc;
    }

    // K0 scaled by 2^wp.
    static BigInteger khinchinFixed(int digits) {
        int prec = (int) Math.ceil((digits + 2) * (Math.log(10) / Math.log(2))) + 32;
        wp = prec + (int) Math.ceil(Math.sqrt(prec)) + 64;
        one = BigInteger.ONE.shiftLeft(wp);
        int bigN = Math.max(3, (int) Math.pow(wp, 0.35));
        int m = (int) Math.ceil(wp * Math.log(2) / (2 * Math.log(bigN))) + 1;

        // -ln((k-1)/k) ln((k+1)/k) = 4 atanh(1/(2k-1)) atanh(1/(2k+1)).
        BigInteger s = BigInteger.ZERO;
        for (long k = 2; k < bigN; k++) {
            s = s.add(mulShift(arcRecip(2 * k - 1, false),
                               arcRecip(2 * k + 1, false)).shiftLeft(2));
        }

        // pi = 16 atan(1/5) - 4 atan(1/239); zeta(2) = pi^2/6 seeds the
        // positive-term recurrence, convolutions via parallel streams.
        BigInteger pi = arcRecip(5, true).shiftLeft(4)
                .subtract(arcRecip(239, true).shiftLeft(2));
        BigInteger[] z = new BigInteger[m + 1];
        z[1] = mulShift(pi, pi).divide(BigInteger.valueOf(6));
        for (int n = 2; n <= m; n++) {
            int half = (n - 1) / 2;
            final int nn = n;
            BigInteger raw;
            if (half >= 64) {
                raw = IntStream.rangeClosed(1, half).parallel()
                        .mapToObj(j -> z[j].multiply(z[nn - j]))
                        .reduce(BigInteger.ZERO, BigInteger::add);
            } else {
                raw = BigInteger.ZERO;
                for (int j = 1; j <= half; j++) {
                    raw = raw.add(z[j].multiply(z[n - j]));
                }
            }
            BigInteger acc = raw.shiftRight(wp).shiftLeft(1);
            if (n % 2 == 0) {
                acc = acc.add(mulShift(z[n / 2], z[n / 2]));
            }
            z[n] = acc.shiftLeft(1).divide(BigInteger.valueOf(2L * n + 1));
        }

        // Main accelerated loop, ascending, incremental power table.
        BigInteger[] powers = new BigInteger[bigN - 2];
        for (int k = 2; k < bigN; k++) {
            powers[k - 2] = one.divide(BigInteger.valueOf((long) k * k));
        }
        BigInteger h = one;
        for (long n = 1; n <= m; n++) {
            BigInteger tail = z[(int) n].subtract(one);
            for (BigInteger p : powers) {
                tail = tail.subtract(p);
            }
            s = s.add(tail.multiply(h).divide(BigInteger.valueOf(n)).shiftRight(wp));
            h = h.add(one.divide(BigInteger.valueOf(2 * n + 1)))
                 .subtract(one.divide(BigInteger.valueOf(2 * n)));
            for (int k = 2; k < bigN; k++) {
                powers[k - 2] = powers[k - 2].divide(BigInteger.valueOf((long) k * k));
            }
        }

        BigInteger ln2 = arcRecip(3, false).shiftLeft(1);
        return expFixed(s.shiftLeft(wp).divide(ln2));
    }

    public static void main(String[] args) throws Exception {
        int digits = args.length > 0 ? Integer.parseInt(args[0]) : 100;
        BigInteger value = khinchinFixed(digits);
        BigInteger scaled = value.multiply(BigInteger.TEN.pow(digits))
                .add(BigInteger.ONE.shiftLeft(wp - 1)).shiftRight(wp);
        String text = scaled.toString();
        String result = text.charAt(0) + "." + text.substring(1);
        if (args.length > 1) {
            Files.writeString(Path.of(args[1]), result + "\n");
        } else {
            System.out.println(result);
        }
    }
}
